File: backend/src/main/java/hu/ibc/ibcpartners/partner/service/ActivityService.java
package hu.ibc.ibcpartners.partner.service;

import hu.ibc.ibcpartners.core.dto.PageResponse;
import hu.ibc.ibcpartners.partner.dto.ActivityDto;
import hu.ibc.ibcpartners.partner.entity.Activity;
import hu.ibc.ibcpartners.partner.mapper.ActivityMapper;
import hu.ibc.ibcpartners.partner.repository.ActivityRepository;
import hu.ibc.ibcpartners.partner.repository.PartnerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
@RequiredArgsConstructor
public class ActivityService {

    private final ActivityRepository activityRepository;
    private final ActivityMapper activityMapper;
    private final PartnerRepository partnerRepository;

    public ActivityDto getById(Long id) {
        return activityMapper.map(findById(id));
    }

    private Activity findById(Long id) {
        return activityRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Tevékenység nem található ezzel az ID-val: " + id));
    }

    public PageResponse<ActivityDto> search(String activity, Pageable pageable) {
        Page<Activity> page = activityRepository.search(activity, pageable);
        return PageResponse.of(page, activityMapper::map);
    }

    public void create(ActivityDto dto) {
        activityRepository.save(activityMapper.map(dto));
    }

    @Transactional
    public void update(ActivityDto dto) {
        Activity activity = findById(dto.id());
        activityMapper.map(dto, activity);
        activityRepository.save(activity);
    }

    public void delete(Long id) {
        if (partnerRepository.activityExists(id)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "A tevékenység nem törölhető, mart használatban van!");
        }
        activityRepository.deleteById(id);
    }
}

File: frontend/src/app/features/core/activity/activity-service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {PageResponse} from '../../page-response';
import {ActivityDto} from './activity-dto';

@Injectable({
  providedIn: 'root',
})
export class ActivityService {
  constructor(private http: HttpClient) {}

  save(req: ActivityDto) {
    if (req.id) {
      return this.http.put(`/api/activities/${req.id}`, req);
    } else {
      return this.http.post('/api/activities', req);
    }
  }

  getById(id: number) {
    return this.http.get<ActivityDto>(`/api/activities/${id}`);
  }

  delete(id: number) {
    return this.http.delete(`/api/activities/${id}`);
  }

  search(req: { page: number; size: number; sort?: string; activity?: string; }) {
    const params: any = {
      page: req.page,
      size: req.size,
      ...(req.sort ? { sort: req.sort } : {}),
      ...(req.activity ? { activity: req.activity } : {}),
    };
    return this.http.get<PageResponse<ActivityDto>>('/api/activities', { params });
  }
}
